### ISSUE TITLE
Support functions that return non-introspected subclasses of introspected types

### ISSUE THREAD

**Issue by romgrk** [romgrk/node-gtk#74]
```js
const gi = require('node-gtk')
const Gtk = gi.require('Gtk', '3.0')
const Gdk = gi.require('Gdk', '3.0')

gi.startLoop()
Gtk.init()
Gdk.init([])


const css = new Gtk.CssProvider()
// css.loadFromPath(path.join(__dirname, 'style.css'))
css.loadFromData(`
button {
  padding: 20px;
}
`.split('').map(s => s.charCodeAt(0)))
const screen = Gdk.Display.getDefaultScreen(Gdk.Display.getDefault())
Gtk.StyleContext.addProviderForScreen(screen, css, 600)
```
Fails with: `gobject.cc: GetClassTemplate: 380: Assertion 'gi_info != NULL' failed: Missing GIR info for: GdkX11Display`

**Comment by romgrk** [comment]
`GdkX11` should be loaded by `Gdk` but is not in it's dependencies.

**Comment by romgrk** [comment]
https://gitlab.gnome.org/GNOME/gobject-introspection/issues/230

**Comment by romgrk** [comment]
Comment from [C. Reiter](https://gitlab.gnome.org/GNOME/gobject-introspection/issues/230#note_316773):

> But note that getDefaultScreen returns a `GdkScreen` subclass, so there is no strict dependency on X11. You can expect functions to return not introspected subclasses of introspected types, e.g. `Gio.File.new_for_path('/')`. You can fall back to the first known base class for example.

Comment from [P. Chimento](https://gitlab.gnome.org/GNOME/gobject-introspection/issues/230#note_317637):
> Sure, here's where we create the class object on demand, in the case where there is no introspection information:
https://gitlab.gnome.org/GNOME/gjs/blob/master/gi/object.cpp#L1771
(The class object is stored in something called the "private namespace object.")
The inheritance mechanism in GJS implements the fallback to the first known base class without us having to do anything other than that.
And, related, here we search through interfaces of a class with no introspection information:
https://gitlab.gnome.org/GNOME/gjs/blob/master/gi/object.cpp#L956

So basically, we can expect functions to return non-introspected subclasses. To deal with it, we can either 1) ind the first known base class, upper in the chain, or 2) create a class object on demand.

**Comment by k0d** [comment]
I’m struggling to understand what the 2 possible solutions are. It seems like they was saying that they create a new class of the known parent type, so it seems like it was just one solution. We could of course create a new class of the non-introspected type which inherits the parents, I guess thats what was meant by a new class. So in the case I am facing with Gstreamer, we could either return a GstPushSrc object, or a GstVideoTestSrc which inherits GstPushSrc.

Maybe I have answered my own question here, what do you think?

I would prefer to return the GstVideoTestSrc class because it might be possible to apply some manual overrides to that class to implement things we need that are not introspectable.

Also, I wonder, why are these not introspectable in the first place? Can we not write manual GIR files to include? I haven’t quite figured out how they are generated and if its for a technicsl reason they can’t be.

**Comment by romgrk** [comment]
My understanding of solution 2 was not clear when I wrote that, maybe indeed there is only solution 1: fall back to the first known base class.

The solution to your particular problem depends on your needs. For example, if you're just using a non-introspected type as a subclass (eg a `GdkX11Screen` as a `GdkScreen`), then solution 1 will fix your issue. However, if you need to use methods that are only defined on the non-introspected subclass, then you will definitely need to introspect that subclass.

GIR files are usually generated from the application source code, using the gnome/gobject [comment documentation format](https://wiki.gnome.org/Projects/GObjectIntrospection/Annotations), as well as reading the source code directly I think (function signature, etc). This is done with the use of the various `g-ir-***` command line tools: `g-ir-scanner`, `g-ir-compiler`, `g-ir-generate`, etc. ([link](https://gi.readthedocs.io/en/latest/tools/index.html))
The correct way to invoke those commands is somewhat mysterious to me, but with some determination you can probably find how to do it. I've never found an actual spec for GIR files, the only spec might be the [code](https://gitlab.gnome.org/GNOME/gobject-introspection). Note that GIR files need their binary counterpart, a .typelib generated with `g-ir-compiler`. I've seen some good tutorials/articles online on how to do it but I don't remember where.

If you want to provide an minimal test case of your problem, it could improve my understanding of this issue overall and would be appreciated. Thanks!

**Comment by k0d** [comment]
Thanks for the info my biggest issue is trying to find out what the methods etc are converted to and only when I am certain that I have found the right JS signature do I know if its a problem with my understanding or the code itself.

This isn’t really a big problem I just have to get deeper into the code. I find all sorts of weird cases and its hard to know why its that way. Your inspect code helps a lot though!!! Maybe I can write some self documenting docs from that.

I also have looked at node-gir and wonder what to suggest...I do think the 2 projects should merge, but exactly how I don’t know. I also wonder if we could make a N-Api version....maybe that will save some headaches in the future. Whatever happens though....I am dedicated to getting this working as it’s important for some other projects :)

**Comment by romgrk** [comment]
Yes, gobject-introspection code is full of weird cases :/

I know about node-gir, and the merge option has already been discussed [extensively](https://github.com/Place1/node-gir/issues/43), but we decided against because we're both too emotionally attached to our own project :) but that's fine, we still help or steal code from each other from time to time.

Regarding n-api, well for the moment Nan solves a subclass of the problem solved by n-api, so it's not a priority (and I also prefer the Nan/V8 API than the n-api). But if there's a popular use-case for it, I'm open to a re-write using n-api.

If you have more issues or questions, don't hesitate to ask.

**Comment by k0d** [comment]
I had read some of the discussion about the possible merge with node-gir, but hadn't noticed if a decision has been made. Because of the memory management in this project I will focus on improving this one, it better fits my requirements :)

N-Api, I don't really know so much about it yet, but I had read good things about it and had tried it out in a little test project and liked it. I guess I'll see how good NaN is soon enough...

I had made some progress yesterday on fixing this issue. I'm back on it now and hopefully will have it sorted by the end of the day.

**Comment by romgrk** [comment]
Sure :) If you want open a PR already with your work in progress, I can monitor and offer advice if you want.

**Comment by k0d** [comment]
It's taking some time to figure things out, but I've got a Gstreamer Pipeline to play now...I have one issue though...

```
var src = Gst.ElementFactory.make("videotestsrc", "src1")
src.setProperty('pattern', 1) // How do I use this...it wants a GValue...not sure how to construct one with the right type
```

**Comment by romgrk** [comment]
Can you provide a full example? I've never used Gstreamer yet. Some kinds of values can be created with GLib (`gi.require('GLib')`), but some might need boxing/unboxing in node-gtk. GValues are handled in node-gtk usually.

**Comment by k0d** [comment]
https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-set-property

I'm trying to set a property on an instance of GstElement which inherits GObject. The method above is what's actually called if I do it in C.

**Comment by romgrk** [comment]
We're using `g_object_set_property` here: [gi.cc:206](https://github.com/romgrk/node-gtk/blob/master/src/gi.cc#L206).

Basically, this will call the `ObjectPropertySetter` function:
```javascript
var src = Gst.ElementFactory.make("videotestsrc", "src1")
src.pattern = 1
```

**Comment by k0d** [comment]
Ok, I see...and I know what my issue is....pattern is a property that's on the GstVideoTestSrc object, which isn't introspectable....so I guess we need some custom 'add property' method on the objects so we can add some in the overrides.

**Comment by romgrk** [comment]
Maybe you can try this:
```js
const internal = require('node-gtk/lib/native.js')
const src = Gst.ElementFactory.make('videotestsrc', 'src1')
internal.ObjectPropertySetter(src, 'pattern', 1)
```

Note that your issue seems a bit different than the issue described here. Unless you open a PR, I'll give solution 1 a try tomorrow.

**Comment by k0d** [comment]
That works...yes...but I'd like to be able to add the property directly to the object. I'll clean up the code and do a PR...
